File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    public void generate(int idx, int target, int[] nums, ArrayList<Integer> vec, List<List<Integer>> res) {
        if (target == 0) {
            res.add(new ArrayList<>(vec));
            return;
        }
        for (int i = idx; i < nums.length; i++) {
            if (i > idx && nums[i] == nums[i - 1]) {
                continue;
            }
            if (nums[i] > target) {
                break;
            }
            vec.add(nums[i]);
            generate(i + 1, target - nums[i], nums, vec, res);
            vec.remove(vec.size() - 1);
        }
    }
    
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(candidates);
        generate(0, target, candidates, new ArrayList<>(), res);
        return res;
    }
}

File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    void f(int i, vector<int>& nums, vector<int>& ds, vector<vector<int>>& res, int target, int& N) {
        if (target == 0) {
            res.push_back(ds);
            return;
        }
        
        for (int j = i; j < N; j++) {
            if (j > i && nums[j] == nums[j - 1]) {
                continue;
            }
            if (nums[j] <= target) {
                ds.push_back(nums[j]);
                f(j + 1, nums, ds, res, target - nums[j], N);
                ds.pop_back();
            }
        }
    }
    
    vector<vector<int>> combinationSum2(vector<int>& nums, int target) {
        int N = nums.size();
        
        vector<vector<int>> res;
        
        vector<int> ds;
        
        sort(nums.begin(), nums.end());
        
        f(0, nums, ds, res, target, N);
        
        return res;
    }
};